native/exe/ServiceControlLaunch/ServiceControlLaunch.cpp

```diff
@@ -45,16 +47,17 @@ JavaVM *vm;
 void sysOutPrint(const char* cstr, ...)
 {
 	JNIEnv* env;
-	jint threadState = vm->GetEnv((void**)&env, JNI_VERSION_1_2);
+	jint threadState = vm->GetEnv((void**)&env, JNI_VERSION_1_6); // JNI in JRE 1.6 and greater
 	if (threadState == JNI_EDETACHED)
 		vm->AttachCurrentThread((void**)&env, NULL);
 	jthrowable oldExcept = env->ExceptionOccurred();
 	if (oldExcept)
 		env->ExceptionClear();
     va_list args;
     va_start(args, cstr);
-    char buf[1024];
-    vsprintf(buf, cstr, args);
+    int const bufsize = 1024;

```

Comment:
This block should be:

char buf[1024];
vsnprintf(buf, sizeof(buf), cstr, args);

That should be portable (not that we need it here) and less lines of code and uses safe string printing.

Reply:
MS docs say vsnprintf() is not safe either, and we don't need to be portable :)
OK to do this instead
    char buf[1024];
    vsprintf_s(buf, sizeof(buf), cstr, args);

Reply:
I'd love to see a pointer to those docs, because that's just wrong unless MS implemented it in an unsafe way. :)  But what you're doing here is fine for windows code.

Reply:
https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/vsnprintf-vsnprintf-vsnprintf-l-vsnwprintf-vsnwprintf-l?view=vs-2015

Reply:
MS is misleading in those docs...but whatever, it's fine with Windows specific calls here.

Reply:
They seem to use the terms "Safe" and "Secure" pretty interchangeably.